app/tests/api/scripts/test_create_user_csv.py

```diff
@@ -1,79 +1,90 @@
-import csv
 import os
+import os.path as path
+import re
 
+import flask.testing
 import pytest
+from pytest_lazyfixture import lazy_fixture
 from smart_open import open as smart_open
 
-from api.services.users.create_user_csv import USER_CSV_RECORD_HEADERS, create_user_csv
-from api.util.file_util import list_files
-from api.util.string_utils import blank_for_null
+import api.adapters.db as db
+import api.app as app_entry
+import tests.api.db.models.factories as factories
+from api.db import models
+from api.db.models.user_models import User
 from tests.api.db.models.factories import UserFactory
+from tests.lib import db_testing
 
 
 @pytest.fixture
-def tmp_file_path(tmp_path):
-    return tmp_path / "example_file.csv"
+def isolated_db(monkeypatch) -> db.DBClient:
```

Comment:
I think there's a separate issue that is at play that needs to be solved first. Each test needs to be independent from each other, otherwise they could conflict when running tests - like you're seeing. If I'm following right, and tinkering with the code right, the rollback doesn't happen between tests, meaning every test is done in a single transaction, rather than independent ones like it did before.

For example, I just pulled up one of the route tests, and the DB objects created from the prior tests are still present. That's going to cause issues (it's caused issues in the past for tests written incorrectly) and means running tests individually and all at once could behave differently.

Before the DB refactor, I believe this was handled by giving the DB session to the app itself, but now the session is fetched from a DB client it generates. There might be a way to get it working that way without any significant dependency injection or refactor:
- Make `create_app` take in an optional db_client (creating it if not supplied)
- The `app` test fixture takes in this `db_client` already, just pass it through
- Create a subclass of `db_client` which overrides the `get_session` method to return the `test_db_session` -might be a few ways to do this, but something to make that session return, perhaps moving the `test_db_session` logic into that method?

Reply:
> Each test needs to be independent from each other, otherwise they could conflict when running tests - like you're seeing.

I agree that tests should be independent, but I think the responsibility should generally be on the test design. In other words, tests generally shouldn't care about what already exists in the database. If a test fails because there was already something in the database from a previous test, I think that's poor test design. I think the case of a service that dumps the entire database to a file is the exception, not the rule.

I actually think the current setup is more isolated than it was previously. Previously, all tests were running under the same transaction, and each test was using a separate nested transaction. I think that unnecessarily prevents the ability of testing code that creates new sessions and new transactions. I think the previous mechanism of having all tests run in a single transaction with nested transactions that automatically rollback is too much "magic", and I prefer a setup that's simpler to understand and forces developers to design their services to be more testable and to adopt simpler/more robust testing practices.

For example, the test_user_route.py tests are still pretty easy to understand and isolated from each other – a test can create a user via POST /users, then fetch/update the user using the uuid of the user. It shouldn't matter if there were existing users in the database or not.

Reply:
(irregardless of the below, does the DB rollback even do anything anymore if the schema gets dropped as you have it setup?)

I agree it comes down to test design, but I've seen several engineers get tripped up by this. Does it make more sense to setup tests in a way that reduces the underlying complexity or reduces the likelihood an engineer makes a test that only works when ran in certain ways?

I'd also really like to emphasize the performance issues that will happen with the isolated_db approach. Mass had that and explicitly moved most tests away from it because it gets incredibly slow to migrate the entire schema for 2k+ tests. I did the switch for most of the ~1k payments tests and brought the runtime down by about 10 minutes.

Is there some other way to do the rollback that is tied to the DB client rather than the nested session?

Reply:
> (irregardless of the below, does the DB rollback even do anything anymore if the schema gets dropped as you have it setup?)

The `test_db_session` (the one that does the rollback) fixture is currently only used by the test_factories tests. I hadn't really gotten around to looking at the test_factories tests to see if we really need to rollback after each test, or if we can redesign the tests to be more agnostic of what's already in the db. If we remove the dependency on `test_db_session` from `test_factories.py`, then we can get rid of `test_db_session` completely. Currently, the fixture does still properly rollback after each test.

> I'd also really like to emphasize the performance issues that will happen with the isolated_db approach. Mass had that and explicitly moved most tests away from it because it gets incredibly slow to migrate the entire schema for 2k+ tests. I did the switch for most of the ~1k payments tests and brought the runtime down by about 10 minutes.

I totally agree. That's why 99% of the tests use the shared schema. The `db_client` fixture in conftest has scope="session" so it creates a single isolated_db for the entire test session.

The `isolated_db` fixture is only used for test_create_user_csv which has the behavior of dumping the entire database contents to a file, which is why that one requires an isolated db. That's the only test that requires an isolated db, everything else uses a shared one and therefore runs pretty fast.

I would actually love to get rid of the `isolated_db` fixture altogether. I think most projects won't even need it. How about we redesign the example backend script create-user-csv to do something more realistic (e.g. dump all users created after a certain time)? If we do that, then we don't need an isolated db to test it.

Reply:
On second thought, maybe it's not unreasonable to assume that some small subset of tests would need an isolated_db. Maybe we could just call it out in docstring/comment that isolated_db is expensive and generally should only be used in situations where you absolutely need an isolated db.

Reply:
It looks like the test_factories.py didn't even really rely on the rollback feature (the tests are already designed in an isolated manner), so I just removed the rollback functionality from `test_db_session` in the last commit since it's not really even being used.

Reply:
I think that sitting down and talking through this would be the best way forward. I consider it a requirement to isolate tests from each other (whether that's rollback or some other mechanism), I know this isn't explicitly due to this PR, but is something I would have pushed back on had I realized it was changed in one of the earlier PRs for the DB session.

Reply:
I think the question is more about what level of isolation rather than whether to be isolated. The previous approach was isolated in some ways and actually less isolated in other ways.

I scheduled a meeting on monday to discuss live, and we can jot down our conversation notes into an ADR for others to see and/or chime in as well.

Reply:
Yes, I think it might be good to go through the types of isolation and what is worth it and what isn't. What types of isolation are worth discussing?
- isolating the DB schema creation has performance issues, so shouldn't be done for the average test
- it's not possible to isolate a DB sequence as they occur outside transactions - you could if made the tests reset sequences between tests, but that's probably more hassle than it's worth
- The DB data itself - isolating helps avoid collisions, but does require the test-specific rollback
- Non-db "static" objects should be either re-generated or otherwise immutable (eg. a base request that other tests override) to avoid one test changing it and that updated version being used
- what else?

Reply:
yeah those sound right.

also, the isolation topic is somewhat closely tied with the topic of levels of abstraction i.e. what abstraction do we want to hide behind the Flask app / test framework, and what we want to expose to underlying application code.

For example, with the previous structure where we created transaction level isolation by starting a transaction and a nested transaction before running the test, it means that the application code can't have it's own concept of transactions. In particular, if the application code tried to do `session.begin()` to start a transaction, it will error with `sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.` If we want to make that work, we'll have to abstract away the concept of transactions from the service methods, and have the Flask application framework handle the creation of the transaction, which means we'll have to force every Flask request to be in exactly one transaction rather than let the service code decide when to start the transaction. I think that's a valid approach, but I worry that the boundaries of any particular "transaction" can be really specific to business logic, so it seems that the service code, rather than the framework, should be the one responsible for deciding when a transaction starts/ends, or whether a transaction is even needed at all.

Reply:
Truncating seems reasonable. It would still add a lot of unnecessary network requests for most tests that are already designed to not care about what was already in the database. Going along that same vein, what if we took the truncation idea and just did it on demand if a particular test needs a clean db? In other words, instead of my "isolated_db" fixture, the create-user-csv test would just truncate the User table at the beginning of the test.